File: Lib/Telerik/Examples/QuickStart/GridView/Filtering/CustomFilterRow/GridFilterExpandableCellElement.cs
using System;
using Telerik.WinControls.UI;
using System.Drawing;

namespace Telerik.Examples.WinControls.GridView.Filtering.CustomFilterRow
{
    public class GridFilterExpandableCellElement : GridRowHeaderCellElement
    {
        #region Fields

        private GridExpanderItem expander;

        #endregion

        #region Initialization

        public GridFilterExpandableCellElement(GridViewColumn col, GridRowElement row)
            : base(col, row)
        {

        }

        protected override void CreateChildElements()
        {
            base.CreateChildElements();

            this.expander = new GridExpanderItem();
            this.expander.SignStyle = SignStyles.Image;
            this.expander.DrawSignBorder = false;
            this.expander.PropertyChanged += new System.ComponentModel.PropertyChangedEventHandler(Expander_PropertyChanged);
            this.Children.Add(expander);
        }

        public override void Initialize(GridViewColumn column, GridRowElement row)
        {
            base.Initialize(column, row);
            this.expander.Expanded = this.RowInfo.IsExpanded;
            SetExpanderImageOrientation(this.expander.Expanded);
        }

        #endregion

        #region Properties

        protected override Type ThemeEffectiveType
        {
            get { return typeof(GridGroupExpanderCellElement); }
        }

        #endregion

        #region Methods

        protected override void UpdateImage()
        {
            this.Image = null;
        }
       
        #endregion

        #region Layout

        protected override SizeF ArrangeOverride(SizeF finalSize)
        {
            base.ArrangeOverride(finalSize);

            RectangleF clientRect = this.GetClientRectangle(finalSize);

            float expanderWidth = this.expander.DesiredSize.Width;
            float imgLeftPos = clientRect.X + (clientRect.Width - expanderWidth) / 2;

            float expanderHeight = this.expander.DesiredSize.Height;
            float imgTopPos = clientRect.Y + (clientRect.Height - expanderHeight) / 2;

            RectangleF imgArea = new RectangleF(imgLeftPos, imgTopPos, expanderWidth, expanderHeight);
            this.expander.Arrange(imgArea);

            return finalSize;
        }

        #endregion

        #region Event Handlers

        private void Expander_PropertyChanged(object sender, System.ComponentModel.PropertyChangedEventArgs e)
        {
            if (this.RowInfo != null)
            {
                this.RowInfo.IsExpanded = expander.Expanded;
            }
        }

        protected override void OnRowPropertyChanged(System.ComponentModel.PropertyChangedEventArgs e)
        {
            base.OnRowPropertyChanged(e);
            this.expander.Expanded = this.RowInfo.IsExpanded;
            SetExpanderImageOrientation(this.expander.Expanded);
        }

        private void SetExpanderImageOrientation(bool isExpanded)
        {
            if (isExpanded)
            {
                this.expander.SignImage = Properties.Resources.chevron_up;
            }
            else
            {
                this.expander.SignImage = Properties.Resources.chevron_down;
            }
        }

        #endregion
    }
}

File: Lib/Telerik/Examples/QuickStart/GridView/Filtering/CustomFilterRow/GridFilterExpandableCellElement.vb
Imports Telerik.WinControls.UI

Namespace Telerik.Examples.WinControls.GridView.Filtering.CustomFilterRow
	Public Class GridFilterExpandableCellElement
		Inherits GridRowHeaderCellElement
		#Region "Fields"

		Private expander As GridExpanderItem

		#End Region

		#Region "Initialization"

		Public Sub New(ByVal col As GridViewColumn, ByVal row As GridRowElement)
			MyBase.New(col, row)

		End Sub

		Protected Overrides Sub CreateChildElements()
			MyBase.CreateChildElements()

			Me.expander = New GridExpanderItem()
			Me.expander.SignStyle = SignStyles.Image
			Me.expander.DrawSignBorder = False
			AddHandler expander.PropertyChanged, AddressOf Expander_PropertyChanged
			Me.Children.Add(expander)
		End Sub

		Public Overrides Sub Initialize(ByVal column As GridViewColumn, ByVal row As GridRowElement)
			MyBase.Initialize(column, row)
			Me.expander.Expanded = Me.RowInfo.IsExpanded
			SetExpanderImageOrientation(Me.expander.Expanded)
		End Sub

		#End Region

		#Region "Properties"

		Protected Overrides ReadOnly Property ThemeEffectiveType() As Type
			Get
				Return GetType(GridGroupExpanderCellElement)
			End Get
		End Property

		#End Region

		#Region "Methods"

		Protected Overrides Sub UpdateImage()
			Me.Image = Nothing
		End Sub

		#End Region

		#Region "Layout"

		Protected Overrides Function ArrangeOverride(ByVal finalSize As SizeF) As SizeF
			MyBase.ArrangeOverride(finalSize)

			Dim clientRect As RectangleF = Me.GetClientRectangle(finalSize)

			Dim expanderWidth As Single = Me.expander.DesiredSize.Width
			Dim imgLeftPos As Single = clientRect.X + (clientRect.Width - expanderWidth) / 2

			Dim expanderHeight As Single = Me.expander.DesiredSize.Height
			Dim imgTopPos As Single = clientRect.Y + (clientRect.Height - expanderHeight) / 2

			Dim imgArea As New RectangleF(imgLeftPos, imgTopPos, expanderWidth, expanderHeight)
			Me.expander.Arrange(imgArea)

			Return finalSize
		End Function

		#End Region

		#Region "Event Handlers"

		Private Sub Expander_PropertyChanged(ByVal sender As Object, ByVal e As System.ComponentModel.PropertyChangedEventArgs)
			If Me.RowInfo IsNot Nothing Then
				Me.RowInfo.IsExpanded = expander.Expanded
			End If
		End Sub

		Protected Overrides Sub OnRowPropertyChanged(ByVal e As System.ComponentModel.PropertyChangedEventArgs)
			MyBase.OnRowPropertyChanged(e)
			Me.expander.Expanded = Me.RowInfo.IsExpanded
			SetExpanderImageOrientation(Me.expander.Expanded)
		End Sub

		Private Sub SetExpanderImageOrientation(ByVal isExpanded As Boolean)
			If isExpanded Then
				Me.expander.SignImage = My.Resources.chevron_up
			Else
				Me.expander.SignImage = My.Resources.chevron_down
			End If
		End Sub

		#End Region
	End Class
End Namespace
